src/data/polynomial/reverse.lean

```diff
@@ -0,0 +1,206 @@
+        rcases (nat.le.dest Nf) with ⟨ F, rfl ⟩,
+        repeat {rw nat.add_sub_cancel_left},
+        repeat {rw rev_at_le},
+        { rw [← add_assoc, add_assoc _ F, add_comm F, ← add_assoc, add_assoc _ F],
+          rw [add_comm f.nat_degree, add_comm F],
+          repeat {rw nat.add_sub_cancel_left}, },
+        { exact le_add_right (le_refl _), },
+        { exact le_add_right (le_refl _), },
+        { rw ← add_assoc (_ + F),
+          rw add_comm _ g.nat_degree,
+          rw ← add_assoc,
+          rw add_assoc,
+          exact nat.le_add_right _ _, }, }, },
+    -- second induction: induction step
+    { intros N O f g Cf Cg Nf Og,
+      by_cases g0 : g = 0,
+      { rw [g0, mul_zero, reflect_zero, reflect_zero, mul_zero], },
+      { rw [← erase_lead_add_C_mul_X_pow g, mul_add, reflect_add, reflect_add, mul_add],
+        rw hcg N O f _ Cf _ Nf (le_trans erase_lead_nat_degree_le Og),
+        rw hcg N O f _ Cf (le_add_left card_support_C_mul_X_pow_le_one) Nf _,
+        { exact (le_trans (nat_degree_C_mul_X_pow_le g.leading_coeff g.nat_degree) Og), },
+        { exact nat.lt_succ_iff.mp (gt_of_ge_of_gt Cg (erase_lead_support_card_lt g0)), }, }, }, },
+  --first induction: induction step
+  { intros N O f g Cf Cg Nf Og,
+    by_cases f0 : f=0,
+    { rw [f0, zero_mul, reflect_zero, reflect_zero, zero_mul], },
+    { rw [← erase_lead_add_C_mul_X_pow f, add_mul, reflect_add, reflect_add, add_mul],
+      rw hcf N O _ g _ Cg (le_trans erase_lead_nat_degree_le Nf) Og,
+      rw hcf N O _ g (le_add_left card_support_C_mul_X_pow_le_one) Cg _ Og,
+      { exact (le_trans (nat_degree_C_mul_X_pow_le f.leading_coeff f.nat_degree) Nf), },
+      { exact nat.lt_succ_iff.mp (gt_of_ge_of_gt Cf (erase_lead_support_card_lt f0)), }, }, },
+end
+
+@[simp] theorem reflect_mul
+  {f g : polynomial R} {F G : ℕ} (Ff : f.nat_degree ≤ F) (Gg : g.nat_degree ≤ G) :
+  reflect (F + G) (f * g) = reflect F f * reflect G g :=
+reflect_mul_induction _ _ F G f g f.support.card.le_succ g.support.card.le_succ Ff Gg
+
+/-- The reverse of a polynomial f is the polynomial obtained by "reading f backwards".
+Even though this is not the actual definition, reverse f = f (1/X) * X ^ f.nat_degree. -/
+noncomputable def reverse (f : polynomial R) : polynomial R := reflect f.nat_degree f
+
+@[simp] lemma reverse_zero : reverse (0 : polynomial R) = 0 := rfl
+
+theorem reverse_mul {f g : polynomial R} (fg : f.leading_coeff * g.leading_coeff ≠ 0) :
+ reverse (f * g) = reverse f * reverse g :=
+begin
+  unfold reverse,
+  rw [nat_degree_mul' fg, reflect_mul (le_refl _) (le_refl _)],
+end
+
+@[simp] lemma reverse_mul_of_domain {R : Type*} [domain R] (f g : polynomial R) :
+  reverse (f * g) = reverse f * reverse g :=
+begin
+  by_cases f0 : f=0,
+  { rw [f0, zero_mul, reverse_zero, zero_mul], },
+  { by_cases g0 : g=0,
+    { rw [g0, mul_zero, reverse_zero, mul_zero], },
+    { apply reverse_mul,
+      apply mul_ne_zero;
+      { rwa [← leading_coeff_eq_zero] at * }, }, },
```

Comment:
You don't necessarily need to make everything go to the right if you finish your first subgoal quickly. Basically in informal mathematics "if `f = 0` then we are done, if `g = 0` then we are done, otherwise ...".

```suggestion
  by_cases f0 : f=0,
  { rw [f0, zero_mul, reverse_zero, zero_mul], },
  by_cases g0 : g=0,
  { rw [g0, mul_zero, reverse_zero, mul_zero], },
  apply reverse_mul,
  apply mul_ne_zero;
    rwa [← leading_coeff_eq_zero] at *
```